torch/_dynamo/variables/nn_module.py

```diff
@@ -414,6 +414,23 @@ def generic_call_method_helper(name):
                 **options,
             )
 
+        if name in ["eval", "train"]:
+            if len(args) == 1:
+                mode = args[0]
+            elif "mode" in kwargs:
+                mode = kwargs.pop("mode")
+            else:
+                assert not (args or kwargs)
+                mode = ConstantVariable.create(name == "train")
+            assert isinstance(mode.as_python_constant(), bool)
+
+            mod = tx.output.get_submodule(self.module_key)
+            mod.training = mode.as_python_constant()
```

Comment:
@ezyang @voznesenskym would this violate “VTs must be immutable”?

Reply:
This is fine IMO. I'd suggest to add a test with graph break, in case there are side effects we need to catch.

Reply:
@yanboliang Why do you think this is fine?

Isn't mod an nn.Module though, not a VT?

Reply:
Basically, I think this needs to be converted into a normal traced setattr/getattr. But I don't know which of the nn.Module properties ought to be guarded.

Reply:
I think that since this is a sideeffect, it actually needs to be turned into a graph break.

However, realistically, we only need to graph break at the _end_ of the graph, i.e. we can use the "delayed_graph_break" strategy, in order to apply the final `setattr`, if any.

Reply:
Note however that while we guard on the top-level nn.Module, I don't know if we guard on the submodules.

Reply:
Actually, I think we do recursively guard on nn.Modules. 

So we can track as a property. We need to recursively check each submodule. If there is a sideeffect (start_value != end_value), we need to emit code to set the attr in python.

Reply:
Thankfully, there already exists infra for this - `SideEffects`. We just need to extend this to `nn.Module`

Reply:
Yes, my point is if we change the nn.module attribute, which is a side effect and should be tracked by dynamo. However, look at the nn.Module reconstruction, it seems fine because every time we load the update module, but I think we should have unit tests for all of these scenarios in case some unexpected changes.

https://github.com/pytorch/pytorch/blob/73170b23d46c7a7569e64ab4e0b6ada5770a2dc9/torch/_dynamo/codegen.py#L134-L143

Reply:
It is OK to track a modification to Module as a side effect, and then replay it in the residual. What is NOT ok is to do what this PR does, which is toggle the attribute *during compilation*.

This is one of the perils of implementing Dynamo directly in Python, it's very easy to accidentally start mutating the runtime env at compilation, but there should always be a strict separation. There isn't any good way to enforce this at runtime, so we must be disciplined in code review.